tests/worker/test_workflow.py

```diff
@@ -2098,6 +2103,74 @@ async def status() -> str:
         assert "never_completing_coroutine" in trace
 
 
+async def test_workflow_enhanced_stack_trace(client: Client):
+    async with new_worker(
+        client, StackTraceWorkflow, LongSleepWorkflow, activities=[wait_cancel]
+    ) as worker:
+        handle = await client.start_workflow(
+            StackTraceWorkflow.run,
+            id=f"workflow-{uuid.uuid4()}",
+            task_queue=worker.task_queue,
+        )
+
+        # Wait until waiting
+        async def status() -> str:
+            return await handle.query(StackTraceWorkflow.status)
+
+        await assert_eq_eventually("waiting", status)
+
+        # Send stack trace query
+        trace = await handle.query("__enhanced_stack_trace")
+
+        assert "never_completing_coroutine" in [
+            loc["functionName"]
+            for stack in trace["stacks"]
+            for loc in stack["locations"]
+        ]
+        # first line of never_completing_coroutine
+        assert 'self._status = "waiting"' in str(trace["sources"])
+        assert trace["sdk"]["version"] == __version__
+
+
+async def test_workflow_external_enhanced_stack_trace(client: Client):
+    async with new_worker(
+        client,
+        ExternalStackTraceWorkflow,
+        ExternalLongSleepWorkflow,
+        activities=[external_wait_cancel],
+    ) as ext_worker:
+        ext_handle = await client.start_workflow(
+            ExternalStackTraceWorkflow.run,
+            id=f"workflow-{uuid.uuid4()}",
+            task_queue=ext_worker.task_queue,
+        )
+
+        async def status() -> str:
+            return await ext_handle.query(ExternalStackTraceWorkflow.status)
+
+        await assert_eq_eventually("waiting", status)
+
+        trace = await ext_handle.query("__enhanced_stack_trace")
+
+        assert "never_completing_coroutine" in [
+            loc["functionName"]
+            for stack in trace["stacks"]
+            for loc in stack["locations"]
+        ]
+
+        fn = None
+        for source in trace["sources"].keys():
+            if source.endswith("externalstacktrace.py 53"):
+                fn = source
```

Comment:
So the thing I was kind of interested to see work is that a workflow defined in one file loading code from _another_ file would properly show both files in the sources.

So, I don't know if you need to copy the whole workflow to the `externalstacktrace` file, but rather just run a coroutine that imports some code from it.

Reply:
gotcha, I'll probably create another file to avoid a circular dependency for it then.